[Merged by Bors] - ci(scripts/*): linting for copyright, imports, module docstrings, line length

bryangingechen opened the pull request:
This PR adds some scripts to check the `.lean` files in `src/` and `archive/` for the following issues (which are out of scope for the current linter):
- Malformed or missing copyright header
- More than one file imported per line
- Module docstring missing, or too late
- Lines of length > 100 (unless they contain `http`)

The scripts are run at the end of our "tests" job since the "lint" job usually takes longer to run. (This isn't a big deal though, since they're quick.)

Current problems are saved in the file `scripts/copy-mod-doc-exceptions.txt` and ignored so that we don't have to fix everything up front. Over time, this should get shorter as we fix things!

Separately, this also fixes some warnings in our GitHub actions workflow (see [this blog post](https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/)).

Co-authored-by: Johan Commelin <johan@commelin.net>

---

We could potentially add a line-length linter to this. Any other ideas? 

Reminder for myself; after this is merged, add a step that calls `scripts/update-copy-mod-doc-exceptions.sh` to the nolints workflow in [azure-scripts](https://github.com/leanprover-community/azure-scripts/blob/master/.github/workflows/nolints.yml).

bryangingechen committed:
chore(*): fix authorship for split files (#4480)

A few files with missing copyright headers in #4477 came from splitting up older files, so the attribution was incorrect:
- `data/setoid/partition.lean` was split from `data/setoid.lean` in #2853.
- `data/finset/order.lean` was split from `algebra/big_operators.lean` in #3495.
- `group_theory/submonoid/operations.lean` was split from `group_theory/submonoid.lean` in  #3058.

jcommelin committed:
chore(scripts/*): crude linting for copyright, imports, module docstrings

bryangingechen committed:
Merge branch 'community-master' into mod-doc-lint

jcommelin committed:
feat(ring_theory/witt_vector/structure_polynomial): witt_structure_rat{_prop} (#4464)

This is the first PR in a series on a fundamental theorem for Witt polynomials.


Co-Authored-By: Rob Y. Lewis <rob.y.lewis@gmail.com>

adomani committed:
feat(data/polynomial/degree/basic): add lemmas dealing with monomials, their support and their nat_degrees (#4475)

bryangingechen committed:
rework scripts and output format

bryangingechen committed:
update github workflow and update_nolints.sh

urkud committed:
feat(measure_theory): integral of a non-negative function is >0 iff μ(support f) > 0 (#4410)

Also add a few supporting lemmas

jsm28 committed:
feat(data/nat/basic): add_succ_lt_add (#4483)

Add the lemma that, for natural numbers, if `a < b` and `c < d` then
`a + c + 1 < b + d` (i.e. a stronger version of `add_lt_add` for the
natural number case).  `library_search` did not find this in mathlib.

hrmacbeth committed:
lint(order/bounded_lattice): docstring (#4484)

fpvandoorn commented on scripts/lint-copy-mod-doc.py:
```diff
@@ -0,0 +1,118 @@
+                errors += [(ERR_COP, line_nr, fn)]
+            if line_nr == copy_start_line_nr + 2 and not "Apache" in line:
+                errors += [(ERR_COP, line_nr, fn)]
+            if line_nr == copy_start_line_nr + 3 and not ("Authors:" in line or "Author:" in line):
```
```suggestion
            if line_nr == copy_start_line_nr + 3 and not "Author" in line:
```
There is also a bunch of `Author(s):`

bryangingechen committed:
allow "Author(s):"

bryangingechen committed:
use enumerate

bryangingechen committed:
update exceptions

awainverse committed:
lint(ring_theory/*): docstrings (#4485)

Docstrings in `ring_theory/ideal/operations`, `ring_theory/multiplicity`, and `ring_theory/ring_invo`.

kmill committed:
feat(control/traversable/basic): composition of applicative transformations (#4487)

Added composition law for applicative transformations, added rest of interface for coercion of applicative transformations to functions (lifted from `monoid_hom`), and proved composition was associative and has an identity.  Also corrected some documentation.

awainverse committed:
lint(order/order_iso_nat): docstrings (#4488)

Adds docstrings to `rel_embedding.nat_lt` and `rel_embedding.nat_gt`

awainverse committed:
lint(order/lexographic, pilex): docstrings (#4489)

Docstrings in `order/lexographic` and `order/pilex`

Commit:
chore(scripts): update nolints.txt (#4490)

I am happy to remove some nolints for you!

bryangingechen committed:
fix enumerate index, relax copyright header check

awainverse committed:
lint(group_theory/perm/*): docstrings (#4492)

Adds missing docstrings in `group_theory/perm/cycles` and `group_theory/perm/sign`.

jcommelin reviewed (approved):
I'm happy with this

jcommelin committed:
Allow comments in import-only files

awainverse committed:
lint(group_theory/*): docstrings and an inhabited instance (#4493)

An inhabited instance for `presented_group`
Docstrings in `group_theory/abelianization` and `group_theory/congruence`.



Co-authored-by: Aaron Anderson <65780815+awainverse@users.noreply.github.com>
Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

jcommelin committed:
Allow unicode copyright symbol in headers

jcommelin committed:
Don't allow empty lines before copyright header

gebner commented on .github/workflows/build.yml:
```diff
@@ -30,8 +30,8 @@ jobs:
           ~/.elan/bin/lean --version
-          echo "::add-path::$HOME/.elan/bin"
-          echo "::set-env name=short_lean_version::$(~/.elan/bin/lean --run scripts/lean_version.lean)"
+          echo "$HOME/.elan/bin" >> $GITHUB_PATH
```
Not unhappy with the change, but it's completely unrelated, right?

jcommelin replied:
Yes, Bryan mentioned a link to a GH blog post about some vulnerability that is fixed by this change.

gebner commented on scripts/lint-copy-mod-doc.py:
```diff
@@ -0,0 +1,122 @@
+with open("scripts/copy-mod-doc-exceptions-short.txt") as f:
+    lines = f.readlines()
+    for line in lines:
+        (fn, c, errno) = line.split()
```
```suggestion
with open("scripts/copy-mod-doc-exceptions.txt") as f:
    for line in f.readlines():
    	fn, _, _, _, _, errno, *_ = l.split()
```

gebner commented on scripts/nolints.txt:
```diff
@@ -7,10 +7,6 @@ apply_nolint category_theory.Kleisli.comp_def unused_arguments
 apply_nolint category_theory.Kleisli.id_def unused_arguments
 apply_nolint category_theory.Kleisli.mk doc_blame has_inhabited_instance
 
--- category_theory/endomorphism.lean
```
Please do not manually change `nolints.txt`

jcommelin replied:
Whoops, merging master fixed this.

gebner commented on src/control/traversable/basic.lean:
```diff
@@ -59,11 +60,12 @@ section applicative_transformation
+`functor.map` (`<*>`) operations. See
+`applicative_transformation.preserves_map` for naturality. -/
 structure applicative_transformation : Type (max (u+1) v w) :=
-(app : ∀ α : Type u, F α → G α)
```
There's lots of unrelated changes in this PR.  Does it depend on some other PRs?

jcommelin replied:
This should be fixed now

jcommelin committed:
Merge branch 'master' into mod-doc-lint

jcommelin commented on scripts/update-copy-mod-doc-exceptions.sh:
```diff
@@ -0,0 +1,11 @@
+
+find src archive -name '*.lean' | xargs ./scripts/lint-copy-mod-doc.py | sort > scripts/copy-mod-doc-exceptions.txt
+
+rm scripts/copy-mod-doc-exceptions-short.txt
```
```suggestion
find src archive -name '*.lean' | xargs ./scripts/lint-copy-mod-doc.py | sort > scripts/copy-mod-doc-exceptions.txt
```

jcommelin committed:
Apply suggestions from code review

Co-authored-by: Gabriel Ebner <gebner@gebner.org>

jcommelin committed:
fix errors

jcommelin committed:
lint for long lines

gebner commented on scripts/update-copy-mod-doc-exceptions.sh:
```diff
@@ -0,0 +1,6 @@
+# This is a little wrapper around the python script
+# lint-copy-mod-doc.py
+
+find src archive -name '*.lean' | xargs ./scripts/lint-copy-mod-doc.py | sort > scripts/copy-mod-doc-exceptions.txt
```
I appreciate the creative use of the `>` operator that makes `lint-copy-mod-doc.py` read its own output, which happens to be empty at first.

jcommelin replied:
Is it too evil?

gebner replied:
It should be be fine.  I don't know if the shell guarantees to open the file before launching the subprocesses, but even if it doesn't `find` takes long enough to prevent the race condition in practice.

bryangingechen replied:
Hmm, it looks like the `>` line got removed. I'd prefer to have that done in this script. I'll open a new PR.

eric-wieser commented on .github/workflows/build.yml:
```diff
@@ -144,3 +144,7 @@ jobs:
           python scripts/yaml_check.py docs/100.yaml docs/overview.yaml docs/undergrad.yaml
           bash scripts/mk_all.sh
           lean --run scripts/yaml_check.lean
+
+      - name: check for copyright headers and module docstrings
+        run: |
```
Can this be run before any of the really slow build steps? I don't want to wait hours for a build, only for it to then say my line is too long and have to start all over again because the file I wrapped the line in needs rebuilding.

bryangingechen replied:
My reasoning was the opposite; since fixing the issues caught by this script shouldn't typically create new build / linter issues, it's better to report those "more serious" problems first so that those can be addressed / discussed. For example, fixing a build / linter issue (e.g. from a github suggestion) might temporarily make a line too long, and I think it'd be better to first check that the suggestion "works" rather than force people to reflow everything up front.

Feel free to bring this up on Zulip, I'd be curious to hear others' thoughts on this as well.

Note that you should be able to run this script manually by executing `./scripts/lint-copy-mod-doc.sh`.
